rustls/src/crypto/signer.rs

```diff
@@ -61,6 +62,12 @@ pub trait SigningKey: Debug + Send + Sync {
 
     /// What kind of key we have.
     fn algorithm(&self) -> SignatureAlgorithm;
+
+    /// Get the RFC 5280-compliant SubjectPublicKeyInfo (SPKI) of this [`SigningKey`] if it exists.
+    fn public_key(&self) -> Result<Option<Vec<u8>>, Error> {
+        // Opt-out by default
+        Err(Error::UnimplementedSubjectPublicKeyInfo)
```

Comment:
I think the branch checked out at this commit won't build because the new `UnimplementedSubjectPublicKeyInfo` error isn't added until the subsequent commit.

Rather than fix that, WDYT about using `Error::General("public_key() unimplemented".into())`? I'm not sure this case merits a dedicated error type.

Reply:
> Rather than fix that, WDYT about using Error::General("public_key() unimplemented".into())? I'm not sure this case merits a dedicated error type.

I like [Ctz's suggestion](https://github.com/rustls/rustls/pull/1954#discussion_r1605018797) to use `Ok(None)` here instead.

Reply:
Shouldn't this be returning SubjectPublicKeyInfoDer?

Reply:
Yes, agreed. Good call-out. Thanks!

Reply:
I updated the return type to `Result<Option<SubjectPublicKeyInfoDer>, Error>` in https://github.com/rustls/rustls/commit/38dd19e0ea9471274a9cf67c06b0aad494966a49. I assume a `pki-types` type is acceptable for a public API in rustls? `cargo check-external-types` isn't working locally for me yet

Reply:
Yup, pki-types are fine in the public API.

Reply:
In case you're interested, the blessed types are specified in the rustls crate `Cargo.toml`: https://github.com/rustls/rustls/blob/c57079361b410bfbdae5f70913d8c89de2e71ed3/rustls/Cargo.toml#L79-L83